Add user guide for Egress feature (#2371)
antrea-io/antrea: docs/egress.md

# Egress

## Table of Contents

<!-- toc -->
- [What is Egress?](#what-is-egress)
- [Prerequisites](#prerequisites)
- [The Egress resource](#the-egress-resource)
  - [AppliedTo](#appliedto)
  - [EgressIP](#egressip)
  - [ExternalIPPool](#externalippool)
- [The ExternalIPPool resource](#the-externalippool-resource)
  - [IPRanges](#ipranges)
  - [nodeSelector](#nodeselector)
- [Usage examples](#usage-examples)
  - [Configuring high available Egress](#configuring-high-available-egress)
  - [Configuring static Egress](#configuring-static-egress)
- [Limitations](#limitations)
<!-- /toc -->

## What is Egress?

`Egress` is a CRD API that manages external access from the Pods in a cluster.
It supports specifying which egress (SNAT) IP the traffic from the selected Pods
to the external network should use. When a selected Pod accesses the external
network, the egress traffic will be tunneled to the Node that hosts the egress
IP if it's different from the Node that the Pod runs on and will be SNATed to
the egress IP when leaving that Node.

> You may be interested in using this capability if any of the below are true:

Reviewer: ```suggestion
You may be interested in using this capability if any of the following apply:
```

Author: done

- A consistent IP address is desired when specific Pods connect to services
  outside of the cluster, for source tracing in audit logs, or whitelisting
  by source IP in external firewall, etc.

> - You want to limit outgoing external connections leave the cluster via certain

Reviewer: ```suggestion
- You want to force outgoing external connections to leave the cluster via certain
```

Author: done

> - You want to limit outgoing external connections leave the cluster via certain

Reviewer 2: ```suggestion
- You want to limit outgoing cluster-external connections via certain
```

Author: Does "limit outgoing cluster-external connections via certain Nodes" have the same meaning? 
What do you think about the latest one: "force outgoing external connections to leave the cluster via certain Nodes"?

>   Nodes, for security controls, or due to the network topology restriction.

Reviewer: or due to network topology restrictions.

Author: done

This guide demonstrates how to configure `Egress` to achieve the above result.

## Prerequisites

Egress is introduced in v1.0 as an alpha feature. As with other alpha features,
a feature gate `Egress` must be enabled on the antrea-controller and
antrea-agent for the feature to work. The following options in the
`antrea-config` ConfigMap need to be set:

```yaml
kind: ConfigMap
apiVersion: v1
metadata:
  name: antrea-config-dcfb6k2hkm
  namespace: kube-system
data:
  antrea-agent.conf: |
    featureGates:
      Egress: true
  antrea-controller.conf: |
    featureGates:
      Egress: true
```

## The Egress resource

A typical Egress resource example:

```yaml
apiVersion: crd.antrea.io/v1alpha2
kind: Egress
metadata:
  name: egress-prod-web
spec:
  appliedTo:
    namespaceSelector:
      matchLabels:
        env: prod
    podSelector:
      matchLabels:
        role: web

>   egressIP: 10.10.0.8

Reviewer: ```suggestion
  egressIP: 10.10.0.8 # can be populated by Antrea after assigning an IP from the pool below
```

Author: done

  externalIPPool: prod-external-ip-pool
status:
  egressNode: node01
```

### AppliedTo

The `appliedTo` field specifies the grouping criteria of Pods to which the
Egress applies to. Pods can be selected cluster-wide using `podSelector`. If set
with a `namespaceSelector`, all Pods from Namespaces selected by the
`namespaceSelector` will be selected. Specific Pods from specific Namespaces can
be selected by providing both a `podSelector` and a `namespaceSelector`. Empty
`appliedTo` selects nothing. The field is mandatory.

### EgressIP

The `egressIP` field specifies the egress (SNAT) IP the traffic from the
selected Pods to the external network should use. **The IP must be reachable

> from all Nodes.** The IP can be specified when creating the Egress. Starting

Reviewer: **The IP must be reachable from all Nodes.**

I wonder if this is really necessary. If I have a cluster with Nodes spread over multiple Node subnets, it should be ok to partition my egress IPs over the different subnets, and each egress IP only needs to be reachable from within the subnet in which it is used. I would configure the `nodeSelector` in my IP pools to only select Nodes within that subnet. But it's probably an unusual scenario and it's easier to frame the requirement as "The IP must be reachable from all Nodes".

Author: The IP must be reachable from all Nodes because we use the egress IPs as tunnel destination (to "encap" the snat IP the traffic should use). As any Pod on any Node can use the egress IPs, the source Node must be able to reach the egress IP, otherwise the tunnel traffic would be dropped.
https://github.com/antrea-io/antrea/blob/main/docs/design/ovs-pipeline.md#snattable-71 has details about the SNAT IP as Tunnel destination IP.

with Antrea v1.2, it can be allocated from an `ExternalIPPool` automatically.

- If `egressIP` is not specified, `externalIPPool` must be specified. An IP will
  be allocated from the pool by the antrea-controller. The IP will be assigned
  to a Node selected by the `nodeSelector` of the `externalIPPool` automatically.
- If both `egressIP` and `externalIPPool` are specified, the IP must be in the
  range of the pool. Similarly, the IP will be assigned to a Node selected by
  the `externalIPPool` automatically.
- If only `egressIP` is specified, Antrea will not manage the assignment of the
  IP and it must be assigned to an arbitrary interface of one Node manually.

**Starting with Antrea v1.2, high availability is provided automatically when
the `egressIP` is allocated from an `externalIPPool`**, i.e. when the
`externalIPPool` is specified. If the Node hosting the `egressIP` fails, another

> Node will be elected as the new egress Node of this Egress. It will take over

Reviewer: suggestion: will be elected (from among the remaining Nodes selected by the `nodeSelector` of the `externalIPPool`) as the ...

Author: done

> the IP and sends layer 2 advertisement (for example, Gratuitous ARP for IPv4) to

Reviewer: s/sends/send

Author: done

associated with the IP has changed.

**Note**: If more than one Egress applies to a Pod and they specify different
`egressIP`, the effective egress IP will be selected randomly.

### ExternalIPPool

The `externalIPPool` field specifies the name of the `ExternalIPPool` that the
`egressIP` should be allocated from. It also determines which Nodes the IP can
be assigned to. It can be empty, which means users should assign the `egressIP`
to one Node manually.

## The ExternalIPPool resource

ExternalIPPool defines one or multiple IP ranges that can be used in the
external network. The IPs in the pool can be allocated to the Egress resources
as the Egress IPs. A typical ExternalIPPool resource example:

```yaml
- apiVersion: crd.antrea.io/v1alpha2
  kind: ExternalIPPool
  metadata:
    name: prod-external-ip-pool
  spec:
    ipRanges:
    - start: 10.10.0.2
      end: 10.10.0.10
    - cidr: 10.10.1.0/28
    nodeSelector:
      matchLabels:
        network-role: egress-gateway
```

### IPRanges

The `ipRanges` field contains a list of IP ranges representing the available IPs
of this IP pool. Each IP range may consist of a `cidr` or a pair of `start` and

> `end` IPs inclusive.

Reviewer: ```suggestion
`end` IPs (which are themselves included in the range).
```

Author: done

### NodeSelector

The `nodeSelector` field specifies which Nodes the IPs in this pool can be
assigned to. It's useful when you want to limit egress traffic to certain Nodes.
The semantics of the selector is the same as those used elsewhere in Kubernetes,

> i.e. both `matchLabel` and `matchExpressions` are supported. It can be empty,

Reviewer 2: ```suggestion
i.e. both `matchLabels` and `matchExpressions` are supported. It can be empty,
```

Author: done

apiVersion: crd.antrea.io/v1alpha2
kind: Egress
metadata:
  name: egress-prod-web
spec:
  appliedTo:
    namespaceSelector:
      matchLabels:
        kubernetes.io/metadata.name: prod
    podSelector:
      matchLabels:
        app: web
  externalIPPool: external-ip-pool
---
apiVersion: crd.antrea.io/v1alpha2
kind: Egress
metadata:
  name: egress-staging-web
spec:
  appliedTo:
    namespaceSelector:
      matchLabels:
        kubernetes.io/metadata.name: staging
    podSelector:
      matchLabels:
        app: web
  externalIPPool: external-ip-pool
```

List the `Egress` resource with kubectl. The output shows each Egress gets one
IP from the IP pool and gets one Node assigned as its Egress Node.

```yaml
# kubectl get egress
NAME                 EGRESSIP       AGE   NODE
egress-prod-web      10.10.0.11     1m    node-4
egress-staging-web   10.10.0.12     1m    node-6
```

Now, the packets from the Pods with label `app=web` in `prod` Namespace to the
external network will be redirected to `node-4` Node and SNATed to `10.10.0.11`
while the packets from the Pods with label `app=web` in `staging` Namespace to
the external network will be redirected to `node-6` Node and SNATed to

> `10.10.0.12`.

Reviewer: in the `prod` Namespaces
to the `node-4` Node
in the `staging` Namespaces
to the `node-6` Node

Author: done. I assume you mean "in the `prod` Namespace", not "Namespace**s**"? I used namespace name label so it can only match one namespace.

Finally, if `node-4` Node powers off, `10.10.0.11` will be re-assigned to

> another available Node in a short time and the packets from the Pods with label

Reviewer: s/in a short time/quickly

Author: done

external network.

Since the Egress IPs have been configured to the Nodes, we can create `Egress`
resources with specific IPs directly.

```yaml
apiVersion: crd.antrea.io/v1alpha2
kind: Egress
metadata:
  name: egress-prod
spec:
  appliedTo:
    namespaceSelector:
      matchLabels:
        kubernetes.io/metadata.name: prod
  egressIP: 10.10.0.104   # node-4's IP
---
apiVersion: crd.antrea.io/v1alpha2
kind: Egress
metadata:
  name: egress-staging
spec:
  appliedTo:
    namespaceSelector:
      matchLabels:
        kubernetes.io/metadata.name: staging
  egressIP: 10.10.0.105   # node-5's IP
```

List the `Egress` resource with kubectl. The output shows `10.10.0.104` is
discovered on `node-4` Node while `10.10.0.105` is discovered on `node-5`.

```yaml
# kubectl get egress
NAME                 EGRESSIP       AGE   NODE
egress-prod          10.10.0.104    1m    node-4
egress-staging       10.10.0.105    1m    node-5
```

Now, the packets from the Pods with in `prod` Namespace to the external network
will be redirected to `node-4` Node and SNATed to `10.10.0.104` while the
packets from the Pods in `staging` Namespace to the external network will be

> redirected to `node-5` Node and SNATed to `10.10.0.12`.

Reviewer: same comment as above

Author: done

to one Node manually.

## The ExternalIPPool resource

ExternalIPPool defines one or multiple IP ranges that can be used in the
external network. The IPs in the pool can be allocated to the Egress resources
as the Egress IPs. A typical ExternalIPPool resource example:

```yaml
- apiVersion: crd.antrea.io/v1alpha2
  kind: ExternalIPPool
  metadata:
    name: prod-external-ip-pool
  spec:
    ipRanges:
    - start: 10.10.0.2
      end: 10.10.0.10
    - cidr: 10.10.1.0/28
    nodeSelector:
      matchLabels:
        network-role: egress-gateway
```

### IPRanges

The `ipRanges` field contains a list of IP ranges representing the available IPs
of this IP pool. Each IP range may consist of a `cidr` or a pair of `start` and
`end` IPs (which are themselves included in the range).

### NodeSelector

The `nodeSelector` field specifies which Nodes the IPs in this pool can be
assigned to. It's useful when you want to limit egress traffic to certain Nodes.
The semantics of the selector is the same as those used elsewhere in Kubernetes,
i.e. both `matchLabels` and `matchExpressions` are supported. It can be empty,
which means all Nodes can be selected.

## Usage examples

> ### Configuring high available Egress

Reviewer 3: I thought it should be "highly available", but after some researching I am not very certain now. @antoninbas : comments?

Personally I prefer "Egress with HA", or "auto-failover Egress".

Reviewer: yes, it should either be "highly-available Egress" or "high-availability Egress"

Author: Thanks, updated to "High-Availability", following https://kubernetes.io/docs/tasks/administer-cluster/highly-available-control-plane
